### DISCUSSION TITLE
Duda sobre excepciones

### DISCUSSION THREAD

**Discussion by AgustinDeCesare** [dds-jv/foro#222] [Consultas de herramientas] [answered]
Hola buenas noches, tenemos una consulta sobre metodos y excepciones.

Si nosotros tenemos un método y este tiene adentro una llamada a otro método que puede fallar y lanzar una excepción, si la excepción se lanza el primer método deja de ejecutarse? O tira la excepción y sigue con el resto del código?

El código es este:

    public void setPassword(String password) throws FileNotFoundException, CreacionError, ErrorPassword {
        // Aca es donde no sabemos si hacer un try catch o si con el que esta incluido en el metodo ya no hace falta
        this.validarPassword();
        this.password = password;
    }

    public abstract Usuario obtenerUsuario();

    public void validarPassword() throws ErrorPassword, FileNotFoundException, CreacionError {
        try{
            // Si la contraseña cumple con lo que tiene que tener
            if(!validador.validarDebilidadPassword(this.password)) throw new ErrorPassword("La password no cumple los requisitos.");

            // Si la contraseña no pertenece al top 10.000
            if(validador.validarTopMalasContrasenias(this.password)) throw new ErrorPassword("La password es demasiado debil. Por favor, elija otra.");
        } catch (ErrorPassword e){
            throw new CreacionError(e);
        }
    }
En este caso si validarPassword tira excepción el set sigue con la asignación o sale?

Gracias

**Comment by flbulgarelli** [comment]
¡Hola!

Las excepciones cortan el flujo de ejecución siempre que no haya un bloque `try` que lo evite. Por lo tanto, sí, si se lanza la excepción en el `if`, la asignación `this.password = password`  en el otro método no se evaluará, y además `setPassword` propagará automáticamente la excepción. 

¡Saludos!

**Comment by flbulgarelli** [comment, marked as answer]
Por otro lado, algunos comentarios sobre el código: 

*   `FileNotFoundException`  es una excepción checkeada, es decir, te obliga a declararla en la firma de cada método que pueda lanzarla. Eso es un problema, y lo que recomiendo es que no la declares y por el contrario la envuelvas en una `RuntimeException` o alguna subclase (este es uno de los pocos casos en que tiene sentido en Java capturar una excepción que de todas formas querías propagar). 
* no le veo mucho sentido a tu `try-catch`  en `validarPassword`, dado que estás cambiando una excepción de dominio por otra, cuando ya `ErrorPassword`  era un buen nombre y aportaba una causa suficientemente específica
* Tu código, tal como está ahora, jamás lanzará ni una `ErrorPassword` , por lo que no es necesario declararla en la firma (de hecho, si hacés que todas tus excepciones hereden de `RuntimeException` jamás necesitarás declararlas. 
* Como la contraseña se está validando antes de que la setees, `validarPassword` va a producir un `NullPointerException` en la mayoría de los casos. 

Con todo esto, te propongo modificar tu código así: 

```java
public void setPassword(String password) {
    this.validarPassword(password);
    this.password = password;
}

public abstract Usuario obtenerUsuario();

public void validarPassword(String password) {
  if(!validador.validarDebilidadPassword(password)) throw new ErrorPassword("La password no cumple los requisitos.");
  if(validador.validarTopMalasContrasenias(password)) throw new ErrorPassword("La password es demasiado debil. Por favor, elija otra.");
}
```

**Reply by gastonprieto** [reply]
Buenas @AgustinDeCesare te sirvio lo que te comento Franco?

**Reply by AgustinDeCesare** [reply]
Si, lo pudimos solucionar con eso. Gracias!
